exchange/events.go

```diff
@@ -0,0 +1,112 @@
+	bidderInfos        adapters.BidderInfos
+	externalURL        string
+}
+
+// getEventTracking creates an eventTracking object from the different configuration sources
+func getEventTracking(requestExtPrebid *openrtb_ext.ExtRequestPrebid, ts time.Time, account *config.Account, bidderInfos adapters.BidderInfos, externalURL string) *eventTracking {
+	return &eventTracking{
+		accountID:          account.ID,
+		enabledForAccount:  account.EventsEnabled,
+		enabledForRequest:  requestExtPrebid != nil && requestExtPrebid.Events != nil,
+		auctionTimestampMs: ts.UnixNano() / 1e+6,
+		integration:        "", // FIXME
+		bidderInfos:        bidderInfos,
+		externalURL:        externalURL,
+	}
+}
+
+// modifyBidsForEvents adds bidEvents and modifies VAST AdM if necessary.
+func (ev *eventTracking) modifyBidsForEvents(seatBids map[openrtb_ext.BidderName]*pbsOrtbSeatBid) map[openrtb_ext.BidderName]*pbsOrtbSeatBid {
+	for bidderName, seatBid := range seatBids {
+		modifyingVastXMLAllowed := ev.isModifyingVASTXMLAllowed(bidderName.String())
+		for _, pbsBid := range seatBid.bids {
+			if modifyingVastXMLAllowed {
+				ev.modifyBidVAST(pbsBid, bidderName)
+			}
+			pbsBid.bidEvents = ev.makeBidExtEvents(pbsBid, bidderName)
+		}
+	}
+	return seatBids
+}
+
+// isModifyingVASTXMLAllowed returns true if this bidder config allows modifying VAST XML for event tracking
+func (ev *eventTracking) isModifyingVASTXMLAllowed(bidderName string) bool {
+	return ev.bidderInfos[bidderName].ModifyingVastXmlAllowed && (ev.enabledForAccount || ev.enabledForRequest)
+}
+
+// modifyBidVAST injects event Impression url if needed, otherwise returns original VAST string
+func (ev *eventTracking) modifyBidVAST(pbsBid *pbsOrtbBid, bidderName openrtb_ext.BidderName) {
+	bid := pbsBid.bid
+	if pbsBid.bidType != openrtb_ext.BidTypeVideo || len(bid.AdM) == 0 && len(bid.NURL) == 0 {
+		return
+	}
+	vastXML := makeVAST(bid)
+	if newVastXML, ok := events.ModifyVastXmlString(ev.externalURL, vastXML, bid.ID, bidderName.String(), ev.accountID, ev.auctionTimestampMs); ok {
+		bid.AdM = newVastXML
+	}
+}
+
+// modifyBidJSON injects "wurl" (win) event url if needed, otherwise returns original json
+func (ev *eventTracking) modifyBidJSON(pbsBid *pbsOrtbBid, bidderName openrtb_ext.BidderName, jsonBytes []byte) []byte {
+	if !ev.enabledForAccount && !ev.enabledForRequest || pbsBid.bidType == openrtb_ext.BidTypeVideo {
+		return jsonBytes
+	}
+	var winEventURL string
+	if pbsBid.bidEvents != nil { // All bids should have already been updated with win/imp event URLs
+		winEventURL = pbsBid.bidEvents.Win
+	} else {
+		winEventURL = ev.makeEventURL(analytics.Win, pbsBid, bidderName)
+	}
+	// wurl attribute is not in the schema, so we have to patch
+	if patch, err := json.Marshal(map[string]string{"wurl": winEventURL}); err == nil {
+		if modifiedJSON, err := jsonpatch.MergePatch(jsonBytes, patch); err == nil {
+			jsonBytes = modifiedJSON
+		}
+	}
+	return jsonBytes
```

Comment:
I thought it shouldn't happen because jsonBytes is valid json (it was marshalled earlier).
But I agree that ignored errors can waste time later, so I can add it to the error list.